src/python/pants/backend/graph_info/tasks/listtargets.py

```diff
@@ -27,14 +26,30 @@ def register_options(cls, register):
     register('--documented', type=bool,
              help='Print only targets that are documented with a description.')
 
+  @classmethod
+  def address_products(cls, options):
+    return (BuildFileAddresses,)
+
   def __init__(self, *args, **kwargs):
     super(ListTargets, self).__init__(*args, **kwargs)
     options = self.get_options()
     self._provides = options.provides
     self._provides_columns = options.provides_columns
     self._documented = options.documented
 
-  def console_output(self, targets):
+  def render(self):
+    
+    visited = set()
+    for addresses in self.get_address_products()[BuildFileAddresses]:
```

Comment:
Hm... that is an interesting thought.

The issue that on the surface would be that we're expecting frontend tasks to be running synchronously in the foreground (ie, post-fork outside of the daemon). So we would not actually be memoizing their execution. The dividing line between the `@rules` and the frontend task body is important in that regard, because it divides between "everything that is parallelized and memoized, which should be almost everything" and "the synchronous foreground bit".

I like the idea of reusing the syntax, but I think that if we did this we'd have to differentiate between `@rule`s (which in theory have zero sideeffects, and no access to the filesystem), and whatever the "foreground task" equivalent is. And we'd have to be crystal clear about the differences in memoization. To make them caceable, we'd need to avoid actually forking in that case too. And I wonder whether our sandboxing is sufficiently powerful to allow for that safely.